### PR TITLE
QUIC p2p

### PR DESCRIPTION
**Pull request by kim**
_Still bit of cleanup and docs required. Look at `examples/mesh.rs` to see actual git cloning in action_

### CODE REVIEW

**Comment by kim** [comment]
I think this is good to go now, as most improvements (imho) are already captured in issues, and I don't want it to grow even bigger.

@FintanH @massimiliano-mantione let me know if you want this to stay open for studying. I'll do all further work in separate branches on top of this, tho.

File: librad/examples/mesh.rs [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+        name: name.into(),
+        key: key.clone(),
+        paths: Paths::from_root(root)?,
+    };
+
+    let git = GitServer {
+        export: peer.paths.projects_dir().into(),
+    };
+
+    let endpoint = Endpoint::bind(&key, "127.0.0.1:0".parse().unwrap()).await?;
```
[the remark this answers is not in the document]
Comment (kim) [reply librad/examples/mesh.rs]: The unwrap is for parsing of the SocketAddr, which we “know” is well-formed. The bind can still fail with an IO error.

Idk, maybe there is a macro somewhere which can validate IP addresses statically, or we can write one? I’m guessing that we will have more than one hard-coded address down the road.

File: librad/examples/mesh.rs [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+    let tmp = tempdir().unwrap();
+    let Bootstrap {
+        peer,
+        proto,
+        endpoint,
+    } = bootstrap(&format!("peer{}", i), device::Key::new(), tmp.path())
+        .await
+        .unwrap();
+
+    let disco = discovery::Static::new(
```
[the remark this answers is not in the document]
Comment (kim) [reply librad/examples/mesh.rs]: I’m still pondering over a combinator for discos, which abides to the law that when you combine two discos you get a new disco with two dance floors.

[the remark this answers is not in the document]
Comment (FintanH) [reply librad/examples/mesh.rs]: We could also have fun things like:
`disco.dancing_at_the_disco_everybody_lets_go_wheres_me_jumper_wheres_me_jumper()`

File: librad/examples/mesh.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (cloudhead) [reply librad/examples/mesh.rs]: Might be possible to use `Ipv4Addr::new(127, 0, 0, 1)`

[the remark this answers is not in the document]
Comment (kim) [reply librad/examples/mesh.rs]: Ah missed that. It's a bit verbose, tho (note the port number)

File: librad/examples/mesh.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (kim) [reply librad/examples/mesh.rs]: Do jumpers form a monoid?

[the remark this answers is not in the document]
Comment (FintanH) [reply librad/examples/mesh.rs]: Monoid in the category of cardigans

**Comment by kim** [comment]
Closed via 294b78597505887740bfa89a1bf1faa034aaad5b

**Pull request closed without merging**